### PR TITLE
[SQServices-1627] Check remaining time of an activated a feature in back office

### PR DESCRIPTION
**Pull request by elland**
The changes to the feature endpoint is backward compatible and shouldn't require any specific deploy order. 

## Checklist

 - [x] The **PR Title** explains the impact of the change.
 - [x] The **PR description** provides context as to why the change should occur and what the code contributes to that effect. This could also be a link to a JIRA ticket or a Github issue, if there is one.
 - [x] **changelog.d** contains the following bits of information ([details](https://github.com/wireapp/wire-server/blob/develop/docs/developer/changelog.md)):
 - [x] If public end-points have been changed or added: does nginz need un upgrade?
 - [x] If internal end-points have been added or changed: which services have to be deployed in a specific order?

### CODE REVIEW

**Comment by fisx** [comment]
did you run the unit tests locally?

```
==== Testing galley-types...
#10 490.3 Tests
#10 490.3   Tests
#10 490.3     owner has all permissions:                                                          OK
#10 490.3     smaller roles (further to the left/top in the type def) are strictly more powerful: OK
#10 490.3     FeatureFlags:                                                                       FAIL (0.01s)
#10 490.3       *** Failed! Falsified (after 1 test):
#10 490.3       FeatureFlags {_flagSSO = FeatureSSODisabledByDefault, _flagLegalHold = FeatureLegalHoldWhitelistTeamsAndImplicitConsent, _flagTeamSearchVisibility = FeatureTeamSearchVisibilityUnavailableByDefault, _flagAppLockDefaults = Defaults {_unDefaults = WithStatusBase {wsbStatus = Identity FeatureStatusEnabled, wsbLockStatus = Identity LockStatusUnlocked, wsbConfig = Identity (AppLockConfig {applockEnforceAppLock = EnforceAppLock True, applockInactivityTimeoutSecs = -1}), wsbTTL = Identity FeatureTTLUnlimited}}, _flagClassifiedDomains = WithStatusBase {wsbStatus = Identity FeatureStatusDisabled, wsbLockStatus = Identity LockStatusUnlocked, wsbConfig = Identity (ClassifiedDomainsConfig {classifiedDomainsDomains = []}), wsbTTL = Identity (FeatureTTLSeconds 0)}, _flagFileSharing = Defaults {_unDefaults = WithStatusBase {wsbStatus = Identity FeatureStatusEnabled, wsbLockStatus = Identity LockStatusUnlocked, wsbConfig = Identity FileSharingConfig, wsbTTL = Identity FeatureTTLUnlimited}}, _flagConferenceCalling = Defaults {_unDefaults = WithStatusBase {wsbStatus = Identity FeatureStatusDisabled, wsbLockStatus = Identity LockStatusUnlocked, wsbConfig = Identity ConferenceCallingConfig, wsbTTL = Identity FeatureTTLUnlimited}}, _flagSelfDeletingMessages = Defaults {_unDefaults = WithStatusBase {wsbStatus = Identity FeatureStatusDisabled, wsbLockStatus = Identity LockStatusUnlocked, wsbConfig = Identity (SelfDeletingMessagesConfig {sdmEnforcedTimeoutSeconds = 0}), wsbTTL = Identity (FeatureTTLSeconds 0)}}, _flagConversationGuestLinks = Defaults {_unDefaults = WithStatusBase {wsbStatus = Identity FeatureStatusEnabled, wsbLockStatus = Identity LockStatusUnlocked, wsbConfig = Identity GuestLinksConfig, wsbTTL = Identity (FeatureTTLSeconds 0)}}, _flagsTeamFeatureValidateSAMLEmailsStatus = Defaults {_unDefaults = WithStatusBase {wsbStatus = Identity FeatureStatusEnabled, wsbLockStatus = Identity LockStatusUnlocked, wsbConfig = Identity ValidateSAMLEmailsConfig, wsbTTL = Identity (FeatureTTLSeconds 0)}}, _flagTeamFeatureSndFactorPasswordChallengeStatus = Defaults {_unDefaults = WithStatusBase {wsbStatus = Identity FeatureStatusDisabled, wsbLockStatus = Identity LockStatusLocked, wsbConfig = Identity SndFactorPasswordChallengeConfig, wsbTTL = Identity (FeatureTTLSeconds 0)}}, _flagTeamFeatureSearchVisibilityInbound = Defaults {_unDefaults = WithStatusBase {wsbStatus = Identity FeatureStatusDisabled, wsbLockStatus = Identity LockStatusUnlocked, wsbConfig = Identity SearchVisibilityInboundConfig, wsbTTL = Identity FeatureTTLUnlimited}}, _flagMLS = Defaults {_unDefaults = WithStatusBase {wsbStatus = Identity FeatureStatusDisabled, wsbLockStatus = Identity LockStatusUnlocked, wsbConfig = Identity (MLSConfig {mlsProtocolToggleUsers = [], mlsDefaultProtocol = ProtocolMLSTag, mlsAllowedCipherSuites = [], mlsDefaultCipherSuite = MLS_128_DHKEMX25519_AES128GCM_SHA256_Ed25519}), wsbTTL = Identity (FeatureTTLSeconds 0)}}}
#10 490.3       Object (fromList [("appLock",Object (fromList [("defaults",Object (fromList [("config",Object (fromList [("enforceAppLock",Bool True),("inactivityTimeoutSecs",Number (-1.0))])),("status",String "enabled"),("ttl",String "unlimited")]))])),("classifiedDomains",Object (fromList [("config",Object (fromList [("domains",Array [])])),("status",String "disabled"),("ttl",String "unlimited")])),("conferenceCalling",Object (fromList [("defaults",Object (fromList [("status",String "disabled"),("ttl",String "unlimited")]))])),("conversationGuestLinks",Object (fromList [("defaults",Object (fromList [("lockStatus",String "unlocked"),("status",String "enabled"),("ttl",String "unlimited")]))])),("fileSharing",Object (fromList [("defaults",Object (fromList [("lockStatus",String "unlocked"),("status",String "enabled"),("ttl",String "unlimited")]))])),("legalhold",String "whitelist-teams-and-implicit-consent"),("mls",Object (fromList [("defaults",Object (fromList [("config",Object (fromList [("allowedCipherSuites",Array []),("defaultCipherSuite",Number 1.0),("defaultProtocol",String "mls"),("protocolToggleUsers",Array [])])),("status",String "disabled"),("ttl",String "unlimited")]))])),("searchVisibilityInbound",Object (fromList [("defaults",Object (fromList [("status",String "disabled"),("ttl",String "unlimited")]))])),("selfDeletingMessages",Object (fromList [("defaults",Object (fromList [("config",Object (fromList [("enforcedTimeoutSeconds",Number 0.0)])),("lockStatus",String "unlocked"),("status",String "disabled"),("ttl",String "unlimited")]))])),("sndFactorPasswordChallenge",Object (fromList [("defaults",Object (fromList [("lockStatus",String "locked"),("status",String "disabled"),("ttl",String "unlimited")]))])),("sso",String "disabled-by-default"),("teamSearchVisibility",String "disabled-by-default"),("validateSAMLEmails",Object (fromList [("defaults",Object (fromList [("status",String "enabled"),("ttl",String "unlimited")]))]))])
#10 490.3       Right (FeatureFlags {_flagSSO = FeatureSSODisabledByDefault, _flagLegalHold = FeatureLegalHoldWhitelistTeamsAndImplicitConsent, _flagTeamSearchVisibility = FeatureTeamSearchVisibilityUnavailableByDefault, _flagAppLockDefaults = Defaults {_unDefaults = WithStatusBase {wsbStatus = Identity FeatureStatusEnabled, wsbLockStatus = Identity LockStatusUnlocked, wsbConfig = Identity (AppLockConfig {applockEnforceAppLock = EnforceAppLock True, applockInactivityTimeoutSecs = -1}), wsbTTL = Identity FeatureTTLUnlimited}}, _flagClassifiedDomains = WithStatusBase {wsbStatus = Identity FeatureStatusDisabled, wsbLockStatus = Identity LockStatusUnlocked, wsbConfig = Identity (ClassifiedDomainsConfig {classifiedDomainsDomains = []}), wsbTTL = Identity (FeatureTTLSeconds 0)}, _flagFileSharing = Defaults {_unDefaults = WithStatusBase {wsbStatus = Identity FeatureStatusEnabled, wsbLockStatus = Identity LockStatusUnlocked, wsbConfig = Identity FileSharingConfig, wsbTTL = Identity FeatureTTLUnlimited}}, _flagConferenceCalling = Defaults {_unDefaults = WithStatusBase {wsbStatus = Identity FeatureStatusDisabled, wsbLockStatus = Identity LockStatusUnlocked, wsbConfig = Identity ConferenceCallingConfig, wsbTTL = Identity FeatureTTLUnlimited}}, _flagSelfDeletingMessages = Defaults {_unDefaults = WithStatusBase {wsbStatus = Identity FeatureStatusDisabled, wsbLockStatus = Identity LockStatusUnlocked, wsbConfig = Identity (SelfDeletingMessagesConfig {sdmEnforcedTimeoutSeconds = 0}), wsbTTL = Identity (FeatureTTLSeconds 0)}}, _flagConversationGuestLinks = Defaults {_unDefaults = WithStatusBase {wsbStatus = Identity FeatureStatusEnabled, wsbLockStatus = Identity LockStatusUnlocked, wsbConfig = Identity GuestLinksConfig, wsbTTL = Identity (FeatureTTLSeconds 0)}}, _flagsTeamFeatureValidateSAMLEmailsStatus = Defaults {_unDefaults = WithStatusBase {wsbStatus = Identity FeatureStatusEnabled, wsbLockStatus = Identity LockStatusUnlocked, wsbConfig = Identity ValidateSAMLEmailsConfig, wsbTTL = Identity (FeatureTTLSeconds 0)}}, _flagTeamFeatureSndFactorPasswordChallengeStatus = Defaults {_unDefaults = WithStatusBase {wsbStatus = Identity FeatureStatusDisabled, wsbLockStatus = Identity LockStatusLocked, wsbConfig = Identity SndFactorPasswordChallengeConfig, wsbTTL = Identity (FeatureTTLSeconds 0)}}, _flagTeamFeatureSearchVisibilityInbound = Defaults {_unDefaults = WithStatusBase {wsbStatus = Identity FeatureStatusDisabled, wsbLockStatus = Identity LockStatusUnlocked, wsbConfig = Identity SearchVisibilityInboundConfig, wsbTTL = Identity FeatureTTLUnlimited}}, _flagMLS = Defaults {_unDefaults = WithStatusBase {wsbStatus = Identity FeatureStatusDisabled, wsbLockStatus = Identity LockStatusUnlocked, wsbConfig = Identity (MLSConfig {mlsProtocolToggleUsers = [], mlsDefaultProtocol = ProtocolMLSTag, mlsAllowedCipherSuites = [], mlsDefaultCipherSuite = MLS_128_DHKEMX25519_AES128GCM_SHA256_Ed25519}), wsbTTL = Identity (FeatureTTLSeconds 0)}}}) /= Right (FeatureFlags {_flagSSO = FeatureSSODisabledByDefault, _flagLegalHold = FeatureLegalHoldWhitelistTeamsAndImplicitConsent, _flagTeamSearchVisibility = FeatureTeamSearchVisibilityUnavailableByDefault, _flagAppLockDefaults = Defaults {_unDefaults = WithStatusBase {wsbStatus = Identity FeatureStatusEnabled, wsbLockStatus = Identity LockStatusUnlocked, wsbConfig = Identity (AppLockConfig {applockEnforceAppLock = EnforceAppLock True, applockInactivityTimeoutSecs = -1}), wsbTTL = Identity FeatureTTLUnlimited}}, _flagClassifiedDomains = WithStatusBase {wsbStatus = Identity FeatureStatusDisabled, wsbLockStatus = Identity LockStatusUnlocked, wsbConfig = Identity (ClassifiedDomainsConfig {classifiedDomainsDomains = []}), wsbTTL = Identity FeatureTTLUnlimited}, _flagFileSharing = Defaults {_unDefaults = WithStatusBase {wsbStatus = Identity FeatureStatusEnabled, wsbLockStatus = Identity LockStatusUnlocked, wsbConfig = Identity FileSharingConfig, wsbTTL = Identity FeatureTTLUnlimited}}, _flagConferenceCalling = Defaults {_unDefaults = WithStatusBase {wsbStatus = Identity FeatureStatusDisabled, wsbLockStatus = Identity LockStatusUnlocked, wsbConfig = Identity ConferenceCallingConfig, wsbTTL = Identity FeatureTTLUnlimited}}, _flagSelfDeletingMessages = Defaults {_unDefaults = WithStatusBase {wsbStatus = Identity FeatureStatusDisabled, wsbLockStatus = Identity LockStatusUnlocked, wsbConfig = Identity (SelfDeletingMessagesConfig {sdmEnforcedTimeoutSeconds = 0}), wsbTTL = Identity FeatureTTLUnlimited}}, _flagConversationGuestLinks = Defaults {_unDefaults = WithStatusBase {wsbStatus = Identity FeatureStatusEnabled, wsbLockStatus = Identity LockStatusUnlocked, wsbConfig = Identity GuestLinksConfig, wsbTTL = Identity FeatureTTLUnlimited}}, _flagsTeamFeatureValidateSAMLEmailsStatus = Defaults {_unDefaults = WithStatusBase {wsbStatus = Identity FeatureStatusEnabled, wsbLockStatus = Identity LockStatusUnlocked, wsbConfig = Identity ValidateSAMLEmailsConfig, wsbTTL = Identity FeatureTTLUnlimited}}, _flagTeamFeatureSndFactorPasswordChallengeStatus = Defaults {_unDefaults = WithStatusBase {wsbStatus = Identity FeatureStatusDisabled, wsbLockStatus = Identity LockStatusLocked, wsbConfig = Identity SndFactorPasswordChallengeConfig, wsbTTL = Identity FeatureTTLUnlimited}}, _flagTeamFeatureSearchVisibilityInbound = Defaults {_unDefaults = WithStatusBase {wsbStatus = Identity FeatureStatusDisabled, wsbLockStatus = Identity LockStatusUnlocked, wsbConfig = Identity SearchVisibilityInboundConfig, wsbTTL = Identity FeatureTTLUnlimited}}, _flagMLS = Defaults {_unDefaults = WithStatusBase {wsbStatus = Identity FeatureStatusDisabled, wsbLockStatus = Identity LockStatusUnlocked, wsbConfig = Identity (MLSConfig {mlsProtocolToggleUsers = [], mlsDefaultProtocol = ProtocolMLSTag, mlsAllowedCipherSuites = [], mlsDefaultCipherSuite = MLS_128_DHKEMX25519_AES128GCM_SHA256_Ed25519}), wsbTTL = Identity FeatureTTLUnlimited}}})
#10 490.3       Use --quickcheck-replay=237033 to reproduce.
#10 490.3       Use -p '/FeatureFlags/' to rerun this test only.
```

File: libs/schema-profunctor/README.md [hunk 1]
```diff
@@ -153,7 +153,7 @@ NonEmpty list, as there are no instances for `ToSchema [a]` or
 `nonEmptyArray` can be used to derive a schema for these if there are
 `ToSchema` instances for `UserId` and `Permission`.
 
-### Sum types
+### Sum types
```
[the remark this answers is not in the document]
Comment (elland) [reply libs/schema-profunctor/README.md:156]: Yes, it usually does a good job. I didn't even spot it myself.

File: libs/schema-profunctor/README.md [hunk 2]
```diff
@@ -203,6 +203,50 @@ Finally, we add a name to the schema using the `named`
...
+              else A.parseFail "Expected ''unlimited' or '0'."
+
+      parseSeconds :: A.Value -> A.Parser FeatureTTL
+      parseSeconds = A.withScientific "FeatureTTL" $
+        \s -> case toBoundedInteger s of
+          Just 0 -> pure FeatureTTLUnlimited
+          Just i -> pure . FeatureTTLSeconds $ i
+          Nothing -> A.parseFail "Expected an integer."
+
+      fromTTL :: FeatureTTL -> Maybe A.Value
```
[the remark this answers is not in the document]
Comment (elland) [reply libs/schema-profunctor/README.md:244]: `mkSchema` expects this function to be a maybe. I imagine b/c Aeson.decode can fail.

[the remark this answers is not in the document]
Comment (fisx) [reply libs/schema-profunctor/README.md:244]: What doesn't make sense to me is this: `fromTTL` goes from the *decoded* value to the json serialization (`Value`).  This is the `encode` operation minus the part from `Value` to `ByteString`, which contains the actually curly brackets and everything.  Nothing should be able to fail there.

I think the fact that you're using `decode` is a mistake.  You don't want to *parse* anything, you want to *render*.  Thereforem my suggestion.

(And now that I look at that again, I wonder if the `fromEnum` is even necessary.  What does `encode @Word` do?  But perhaps the fact that I'm asking myself this question is a good reason to keep the `fromEnum`, because it makes the question unnecessary.)

File: Makefile [hunk 3]
```diff
@@ -38,6 +38,10 @@ default: fast
 init:
 	mkdir -p dist
 
+.PHONY: generate-cabal-from-hpack
+generate-cabal-from-hpack:
+	stack build --dry-run
```
Comment (fisx) [inline Makefile:43]: Are you sure you need a make goal for this?  Seems like one of those private tricks that will get lost and just clutter the Makefile.  (I don't insist on removing it, your call.)

File: libs/types-common/src/Util/Options.hs [hunk 4]
```diff
@@ -129,7 +129,11 @@ getOptions desc pars defaultPath = do
     (True, _) -> do
       configFile <- decodeFileEither path
       case configFile of
-        Left e -> fail $ show e
+        Left e ->
+          fail $
+            show e
+              <> " while attempting to decode "
+              <> show path
```
Comment (fisx) [inline libs/types-common/src/Util/Options.hs:136]: Great change, but please put something like this into a separate PR next time.

File: libs/schema-profunctor/README.md (hunk 1 above)
Comment (fisx) [inline libs/schema-profunctor/README.md:156]: this hasn't been fixed yet.  was i supposed to review this again?  i'll stop now!  :)

File: libs/wire-api/src/Wire/API/Team/Feature.hs [hunk 5]
```diff
@@ -317,6 +326,7 @@ instance (ToSchema cfg, IsFeatureConfig cfg) => ToSchema (WithStatusNoLock cfg)
       WithStatusNoLock
         <$> wssStatus .= field "status" schema
         <*> wssConfig .= objectSchema @cfg
+        <*> wssTTL .= field "ttl" schema
```
Comment (fisx) [inline libs/wire-api/src/Wire/API/Team/Feature.hs:329]:
```suggestion
        <*> wssTTL .= fromMaybe FeatureTTLUnlimited <$> field "ttl" schema
```

If you this this, you can drop it from the `values.yaml` above, which I really want to see (so many places where that would have to be kept in mind, and so many people involved, and it really doesn't belong there in the first place.)

**Review by fisx — CHANGES_REQUESTED** [review]

File: Makefile (hunk 3 above)
Comment (elland) [reply Makefile:43]: It's a private trick without which adding a new dependency to package.yaml doesn't work. We seem to have a lot of things but no single source of truth about them, like the docker ephemeral thing.

File: services/brig/test/integration/API/Internal.hs [hunk 6]
```diff
@@ -217,8 +217,8 @@ testFeatureConferenceCallingByAccount (Opt.optSettings -> settings) mgr db brig
           . (^?! Aeson.key "conferenceCalling")
           . responseJsonUnsafe @Aeson.Value
 
-  check $ ApiFt.WithStatusNoLock ApiFt.FeatureStatusEnabled ApiFt.ConferenceCallingConfig
-  check $ ApiFt.WithStatusNoLock ApiFt.FeatureStatusDisabled ApiFt.ConferenceCallingConfig
+  check $ ApiFt.WithStatusNoLock ApiFt.FeatureStatusEnabled ApiFt.ConferenceCallingConfig ApiFt.FeatureTTLUnlimited
+  check $ ApiFt.WithStatusNoLock ApiFt.FeatureStatusDisabled ApiFt.ConferenceCallingConfig ApiFt.FeatureTTLUnlimited
```
[the remark this answers is not in the document]
Comment (elland) [reply services/brig/test/integration/API/Internal.hs:222]: @fisx on top of the integration tests that check TTL expiration results in the flag being reverted back to default? 🤔 

Also since we changed the record used for the requests, we can indeed use it via PUT, so I've started simplifying that on the receiving end of stern, so we no longer need to be passing the TTL around as an extra parameter 👍

File: libs/wire-api/src/Wire/API/Team/Feature.hs [hunk 7]
```diff
@@ -344,7 +354,41 @@ data FeatureTTL
...
+        A.withText "FeatureTTL" $
+          \t ->
+            if t == "unlimited" || t == "0"
+              then pure FeatureTTLUnlimited
+              else A.parseFail "Expected ''unlimited' or '0'."
+
+      parseSeconds :: A.Value -> A.Parser FeatureTTL
+      parseSeconds = A.withScientific "FeatureTTL" $
+        \s -> case toBoundedInteger s of
+          Just 0 -> A.parseFail "Expected a positive integer. Got 0 instead."
```
Comment (fisx) [inline libs/wire-api/src/Wire/API/Team/Feature.hs:384]:
```suggestion
          Just 0 -> error "impossible (this would have parsed in `parseUnlimited` above)."
```

Otherwise I think this is fine.  (I said I wanted to look at this un-isomorphismish-ness again, but I think we should leave it as is.)

File: services/brig/test/integration/API/Internal.hs (hunk 6 above)
[the remark this answers is not in the document]
Comment (fisx) [reply services/brig/test/integration/API/Internal.hs:222]: > @fisx on top of the integration tests that check TTL expiration results in the flag being reverted back to default? thinking

no, if you have a test for that i'm happy, must have missed that.

File: Makefile (hunk 3 above)
Comment (fisx) [reply Makefile:43]: I'll remove this and make a separate PR that gets rid of the package.yaml files.  I remember talking about this in the past, and I think there was something close to consensus about that approach.  Let's see if it passes review!  :)

**Comment by fisx** [comment]
### TODO

- [ ] `( curl -XPUT -H'content-type: application/json' -H'Z-User: e8920a12-289c-47e9-92aa-724d50985d76' http://localhost:8085/i/teams/2af1c39b-9833-4330-bf12-6827cc77e9ba/features/conferenceCalling -d '{"lockStatus":"unlocked","status":"disabled"}' ; curl -XPUT -H'content-type: application/json' -H'Z-User: e8920a12-289c-47e9-92aa-724d50985d76' http://localhost:8085/i/teams/2af1c39b-9833-4330-bf12-6827cc77e9ba/features/conferenceCalling -d '{"lockStatus":"unlocked","status":"enabled","ttl":3}' ) | jq .` prints out `"unlimited"` in both calls.  the second should print `3`.  (In case you're wondering: also works with larger numbers, and the query param also doesn't work any more, and it shouldn't.)
- [ ] test stern, add screen shots here.
- [ ] drop limitation of TTL selection in stern to `conferenceCalling`.  makes code easier and more powerful.
- [ ] limit days in stern to 2000 (galley complains for 7200).
- [ ] conversion from ttl days to seconds needs a phantom type parameter.

File: services/galley/src/Galley/Cassandra/TeamFeatures.hs [hunk 8]
```diff
@@ -238,27 +227,49 @@ instance FeatureStatusCassandra SelfDeletingMessagesConfig where
...
+        fromString $
+          "select conference_calling, ttl(conference_calling) from team_features where team_id = ?"
+
+  setFeatureConfig _ tid statusNoLock =
+    retry x5 $ write insert (params LocalQuorum (tid, wssStatus statusNoLock))
+    where
+      renderFeatureTtl :: FeatureTTL -> String
+      renderFeatureTtl = \case
+        FeatureTTLSeconds d | d > 0 -> " using ttl " <> show d
+        _ -> " using ttl 0" -- 0 or unlimited (delete a column's existing TTL by setting its value to zero)
```
Comment (fisx) [inline services/galley/src/Galley/Cassandra/TeamFeatures.hs:248]: ah, i see.  i got that wrong earlier, thanks!

**Review by fisx — APPROVED** [review]

**Comment by battermann** [comment]
Screenshot from 2022-08-15 14-29-06
Screenshot from 2022-08-15 14-30-19
Screenshot from 2022-08-15 14-30-49
Screenshot from 2022-08-15 14-31-16
